riscv/src/continuations/memory_merkle_tree.rs

```diff
@@ -0,0 +1,272 @@
+use std::collections::BTreeMap;
+
+use crate::bootloader::{
+    BYTES_PER_WORD, MERKLE_TREE_DEPTH as MERKLE_TREE_DEPTH_BOOTLOADER,
+    WORDS_PER_PAGE as WORDS_PER_PAGE_BOOTLOADER,
+};
+use number::FieldElement;
+use riscv_executor::poseidon_gl::poseidon_gl;
+
+/// A Merkle tree of memory pages.
+pub struct MerkleTree<
+    T: FieldElement,
+    const MERKLE_TREE_DEPTH: usize = MERKLE_TREE_DEPTH_BOOTLOADER,
+    const WORDS_PER_PAGE: usize = WORDS_PER_PAGE_BOOTLOADER,
+> {
+    /// Hashes of all nodes of the Merkle tree. The vector has MERKLE_TREE_DEPTH + 1 entries
+    /// where hashes[0] is [root_hash], hashes[1] is [level_1_hash_0, level_1_hash_1], etc.
+    /// The last entry hashes[MERKLE_TREE_DEPTH] is the hash of the leaves.
+    /// These hashes should be updated whenever the data is updated.
+    hashes: Vec<Vec<[T; 4]>>,
+    /// Memory pages.
+    data: Vec<[T; WORDS_PER_PAGE]>,
+}
+
+/// Computes the Poseidon hash of two 4-field-element inputs, using a capacity of 0.
+fn hash_cap0<T: FieldElement>(data1: &[T; 4], data2: &[T; 4]) -> [T; 4] {
+    let mut buffer = [T::zero(); 12];
+    buffer[..4].copy_from_slice(data1);
+    buffer[4..8].copy_from_slice(data2);
+    poseidon_gl(&buffer)
+}
+
+impl<T: FieldElement, const MERKLE_TREE_DEPTH: usize, const WORDS_PER_PAGE: usize>
+    MerkleTree<T, MERKLE_TREE_DEPTH, WORDS_PER_PAGE>
+{
+    /// Build a new Merkle tree starting from an all-zero memory.
+    pub fn new() -> Self {
+        let zero_page = [T::zero(); WORDS_PER_PAGE];
+        let mut hash = Self::hash_page(&zero_page);
+        let mut hashes = vec![vec![]; MERKLE_TREE_DEPTH + 1];
+
+        for level in (0..=MERKLE_TREE_DEPTH).rev() {
+            hashes[level] = vec![hash; 1 << level];
+            hash = hash_cap0(&hash, &hash);
+        }
+
+        assert_eq!(hashes[0].len(), 1);
+
```

Comment:
I'm actually more concerned about `level` being too large and thus the `1<<level` shifting too much to the left. Could you check `level` against the number of bits in `usize`?